In [120]:
import pandas as pd

campaign = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/bank.csv')

In [121]:
# helper functions
def data_inspections(field):
  # print out the describe, value counts of the given field.
  print("\n")
  print("Describe\n",campaign[field].describe())
  print("\n")
  print("Value Counts\n",campaign[field].value_counts())
  print("\n")

  # This list is the list of values to check for as if they represent null.
  null_value_check_list = ['nonexistent', 'unknown']
  for nv in null_value_check_list:
  # check if null values exist and if so, show the percent in of them that are in the column.
    if nv in campaign[field].value_counts():
      print(f"{nv} percent")
      print(campaign[field].value_counts()[nv] / campaign[field].count() * 100)
      print("\n")

  # print actual null value counts
  print("The actual nulls count:",campaign[field].isna().sum())

# Data inspection


## Age



In [122]:
data_inspections('age')



Describe
 count    37069.000000
mean        40.025493
std         10.435288
min         17.000000
25%         32.000000
50%         38.000000
75%         47.000000
max         98.000000
Name: age, dtype: float64


Value Counts
 age
31    1748
32    1658
33    1640
36    1594
34    1581
      ... 
91       2
98       2
95       1
87       1
94       1
Name: count, Length: 78, dtype: int64


The actual nulls count: 0


## Job


In [123]:
data_inspections('job')



Describe
 count      37069
unique        12
top       admin.
freq        9400
Name: job, dtype: object


Value Counts
 job
admin.           9400
blue-collar      8329
technician       6066
services         3548
management       2648
retired          1534
entrepreneur     1308
self-employed    1274
housemaid         962
unemployed        913
student           793
unknown           294
Name: count, dtype: int64


unknown percent
0.7931155412878685


The actual nulls count: 0


## Marital

In [124]:
data_inspections('marital')



Describe
 count       37069
unique          4
top       married
freq        22456
Name: marital, dtype: object


Value Counts
 marital
married     22456
single      10410
divorced     4134
unknown        69
Name: count, dtype: int64


unknown percent
0.1861393617308263


The actual nulls count: 0


## Education

In [125]:
data_inspections('education')



Describe
 count                 37069
unique                    8
top       university.degree
freq                  10949
Name: education, dtype: object


Value Counts
 education
university.degree      10949
high.school             8554
basic.9y                5427
professional.course     4742
basic.4y                3764
basic.6y                2082
unknown                 1535
illiterate                16
Name: count, dtype: int64


unknown percent
4.1409263805336


The actual nulls count: 0


## Default

In [126]:
data_inspections("default")



Describe
 count     37069
unique        3
top          no
freq      29341
Name: default, dtype: object


Value Counts
 default
no         29341
unknown     7725
yes            3
Name: count, dtype: int64


unknown percent
20.83951549812512


The actual nulls count: 0


## Housing

In [127]:
data_inspections("housing")



Describe
 count     37069
unique        3
top         yes
freq      19474
Name: housing, dtype: object


Value Counts
 housing
yes        19474
no         16701
unknown      894
Name: count, dtype: int64


unknown percent
2.4117186867733147


The actual nulls count: 0


## Loan

In [128]:
data_inspections("loan")



Describe
 count     37069
unique        3
top          no
freq      30590
Name: loan, dtype: object


Value Counts
 loan
no         30590
yes         5585
unknown      894
Name: count, dtype: int64


unknown percent
2.4117186867733147


The actual nulls count: 0


# Last contact info

## Contact

In [129]:
data_inspections("contact")



Describe
 count        37069
unique           2
top       cellular
freq         23515
Name: contact, dtype: object


Value Counts
 contact
cellular     23515
telephone    13554
Name: count, dtype: int64


The actual nulls count: 0


## month

In [130]:
data_inspections("month")



Describe
 count     37069
unique       10
top         may
freq      12370
Name: month, dtype: object


Value Counts
 month
may    12370
jul     6445
aug     5555
jun     4817
nov     3698
apr     2369
oct      653
sep      508
mar      496
dec      158
Name: count, dtype: int64


The actual nulls count: 0


## Day of the week

In [131]:
data_inspections('day_of_week')



Describe
 count     37069
unique        5
top         thu
freq       7742
Name: day_of_week, dtype: object


Value Counts
 day_of_week
thu    7742
mon    7657
wed    7347
tue    7287
fri    7036
Name: count, dtype: int64


The actual nulls count: 0


# Other Attributes

## Campaign
The column

In [132]:
data_inspections('campaign')



Describe
 count    37069.000000
mean         2.564407
std          2.764084
min          1.000000
25%          1.000000
50%          2.000000
75%          3.000000
max         56.000000
Name: campaign, dtype: float64


Value Counts
 campaign
1     15874
2      9554
3      4804
4      2378
5      1432
6       855
7       559
8       363
9       265
10      207
11      164
12      112
13       81
14       55
17       50
15       47
16       44
18       30
20       29
19       26
21       21
22       17
23       13
24       13
27       10
29       10
25        7
31        7
30        7
28        7
26        6
35        5
33        4
32        3
40        2
34        2
42        2
39        1
56        1
41        1
37        1
Name: count, dtype: int64


The actual nulls count: 0


## pdays

In [133]:
data_inspections('pdays')



Describe
 count    37069.000000
mean       962.221803
std        187.531477
min          0.000000
25%        999.000000
50%        999.000000
75%        999.000000
max        999.000000
Name: pdays, dtype: float64


Value Counts
 pdays
999    35696
3        396
6        371
4        107
7         56
9         56
12        55
2         54
10        50
5         42
13        32
11        26
1         23
15        22
14        18
8         16
0         14
16        10
17         8
18         6
19         3
21         2
22         2
26         1
25         1
27         1
20         1
Name: count, dtype: int64


The actual nulls count: 0


## previous

In [134]:
data_inspections('previous')



Describe
 count    37069.000000
mean         0.173730
std          0.496159
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          7.000000
Name: previous, dtype: float64


Value Counts
 previous
0    31988
1     4116
2      684
3      196
4       64
5       15
6        5
7        1
Name: count, dtype: int64


The actual nulls count: 0


## poutcome
poutcome - nonexisten is > 60%

We can see that 31988 entries are nonexistent

In [135]:
data_inspections('poutcome')



Describe
 count           37069
unique              3
top       nonexistent
freq            31988
Name: poutcome, dtype: object


Value Counts
 poutcome
nonexistent    31988
failure         3843
success         1238
Name: count, dtype: int64


nonexistent percent
86.29312902964742


The actual nulls count: 0


# Social and Economic Context Attribs

## emp.var.rate

In [136]:
data_inspections("emp.var.rate")



Describe
 count    37069.000000
mean         0.081526
std          1.572287
min         -3.400000
25%         -1.800000
50%          1.100000
75%          1.400000
max          1.400000
Name: emp.var.rate, dtype: float64


Value Counts
 emp.var.rate
 1.4    14605
-1.8     8234
 1.1     7001
-0.1     3314
-2.9     1520
-3.4      973
-1.7      692
-1.1      572
-3.0      150
-0.2        8
Name: count, dtype: int64


The actual nulls count: 0


## cons.price.idx

In [137]:
data_inspections("cons.price.idx")



Describe
 count    37069.000000
mean        93.576551
std          0.579339
min         92.201000
25%         93.075000
50%         93.749000
75%         93.994000
max         94.767000
Name: cons.price.idx, dtype: float64


Value Counts
 cons.price.idx
93.994    7001
93.918    6000
92.893    5171
93.444    4647
94.465    3958
93.200    3250
93.075    2213
92.201     696
92.963     661
92.431     400
92.649     332
94.215     282
94.199     267
92.843     256
92.379     241
93.369     240
94.027     212
93.876     198
94.055     198
94.601     189
92.469     163
93.749     156
92.713     150
94.767     116
93.798      64
92.756       8
Name: count, dtype: int64


The actual nulls count: 0


## cons.conf.idx

In [138]:
data_inspections("cons.conf.idx")



Describe
 count    37069.000000
mean       -40.494829
std          4.628895
min        -50.800000
25%        -42.700000
50%        -41.800000
75%        -36.400000
max        -26.900000
Name: cons.conf.idx, dtype: float64


Value Counts
 cons.conf.idx
-36.4    7001
-42.7    6000
-46.2    5171
-36.1    4647
-41.8    3958
-42.0    3250
-47.1    2213
-31.4     696
-40.8     661
-26.9     400
-30.1     332
-40.3     282
-37.5     267
-50.0     256
-29.8     241
-34.8     240
-38.3     212
-40.0     198
-39.8     198
-49.5     189
-33.6     163
-34.6     156
-33.0     150
-50.8     116
-40.4      64
-45.9       8
Name: count, dtype: int64


The actual nulls count: 0


## euribor3m

In [139]:
data_inspections("euribor3m")



Describe
 count    37069.000000
mean         3.621945
std          1.734496
min          0.634000
25%          1.344000
50%          4.857000
75%          4.961000
max          5.045000
Name: euribor3m, dtype: float64


Value Counts
 euribor3m
4.857    2578
4.962    2346
4.963    2222
4.961    1728
4.856    1103
         ... 
3.488       1
3.743       1
3.816       1
0.956       1
0.953       1
Name: count, Length: 314, dtype: int64


The actual nulls count: 0


## nr.employed

In [140]:
data_inspections('nr.employed')



Describe
 count    37069.000000
mean      5167.010650
std         72.294476
min       4963.600000
25%       5099.100000
50%       5191.000000
75%       5228.100000
max       5228.100000
Name: nr.employed, dtype: float64


Value Counts
 nr.employed
5228.1    14605
5099.1     7640
5191.0     7001
5195.8     3314
5076.2     1520
5017.5      973
4991.6      692
5008.7      594
4963.6      572
5023.5      150
5176.3        8
Name: count, dtype: int64


The actual nulls count: 0


# Output Variable

## "Y" field

In [141]:
data_inspections('y')



Describe
 count     37069
unique        2
top          no
freq      32861
Name: y, dtype: object


Value Counts
 y
no     32861
yes     4208
Name: count, dtype: int64


The actual nulls count: 0
